File: 2025/08 August/31 Container With Most Water/main.py
# Link: https://www.geeksforgeeks.org/problems/container-with-most-water0535/1


class Solution:

    def maxWater(self, arr):
        left = 0
        right = len(arr) - 1
        res = 0
        while left < right:
            water = min(arr[left], arr[right]) * (right - left)
            res = max(res, water)

            if arr[left] < arr[right]:
                left += 1
            else:
                right -= 1

        return res


if __name__ == "__main__":
    sol = Solution()
    arr = [1, 5, 4, 3]
    result = sol.maxWater(arr)
    print("Maximum water that can be contained is:", result)

    arr = [3, 1, 2, 4, 5]
    result = sol.maxWater(arr)
    print("Maximum water that can be contained is:", result)

    arr = [2, 1, 8, 6, 4, 6, 5, 5]
    result = sol.maxWater(arr)
    print("Maximum water that can be contained is:", result)

File: 2025/08 August/31 Container With Most Water/main.cpp
// Link: https://www.geeksforgeeks.org/problems/container-with-most-water0535/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int maxWater(vector<int> &arr)
    {
        int left = 0, right = arr.size() - 1;
        int res = 0;
        while (left < right)
        {
            int water = min(arr[left], arr[right]) * (right - left);
            res = max(res, water);

            if (arr[left] < arr[right])
                left += 1;
            else
                right -= 1;
        }

        return res;
    }
};

int main()
{
    Solution sol;
    vector<int> arr = {1, 5, 4, 3};
    int result = sol.maxWater(arr);
    cout << "Maximum water that can be contained: " << result << endl;

    arr = {3, 1, 2, 4, 5};
    result = sol.maxWater(arr);
    cout << "Maximum water that can be contained: " << result << endl;

    arr = {2, 1, 8, 6, 4, 6, 5, 5};
    result = sol.maxWater(arr);
    cout << "Maximum water that can be contained: " << result << endl;

    return 0;
}